src/Services/CachedServiceAccountJwtRetriever.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace KeycloakAuthGuard\Services;
+
+use Illuminate\Cache\Repository;
+use Illuminate\Http\Client\RequestException;
+use Illuminate\Support\Facades\Config;
+use Psr\SimpleCache\InvalidArgumentException;
+
+readonly class CachedServiceAccountJwtRetriever implements ServiceAccountJwtRetrieverInterface
+{
+    private string $realm;
+
+    public function __construct(private ServiceAccountJwtRetriever $jwtRetriever, private Repository $repository)
+    {
+        $this->realm = Config::get('keycloak.realm');
+    }
+
+    /**
+     * @throws RequestException
+     * @throws InvalidArgumentException
+     */
+    public function getJwt(): string
+    {
+        if ($this->repository->has($this->getCacheKey())) {
+            return $this->repository->get($this->getCacheKey());
+        }
+
+        $jwtResponse = $this->jwtRetriever->getResponse();
+        $this->repository->set(
+            $this->getCacheKey(),
+            $jwtResponse['access_token'],
+            $jwtResponse['expires_in']
+        );
+
+        return $jwtResponse['access_token'];
+    }
```

Comment:
**[blocker]** Even though `$jwtResponse['expires_in']` is passed to `Repository::set`, I think it’s still possible that this method might return an expired access token (causing failure when decoding JWT). 

I suspect this because there is a delay between (1) when the Keycloak server issues the access token and (2) when our cache is updated (here). Since Keycloak’s default access token expiration time is 5 minutes, I think the likelihood of this issue occurring every once in a while is too big to ignore.

Some ideas on how to avoid this issue: 
* Set the cache item’s TTL as slightly less than the "expires_in" value from the response (e.g. `$jwtResponse['expires_in'] - 10`).
* Parse the JWT string to JSON in order to retrieve the `"exp"` value from the payload.
* Run the `\Firebase\JWT\JWT::decode` function on the JWT string, rewuesting a new JWT if `\Firebase\JWT\ExpiredException` is thrown.

Reply:
Usually it is a good idea to react to the expiring token before the expiry - it will leave some room to request new token with the old token still working. This time can be used when some issues occur and new token needs to be queried etc. E.g. then token TTL is 5 min, then start requesting at 4 min or when 10 min then 2 min or something like that.

The expires_in field in Keycloak's token endpoint response should be the same as in token, but to add more consistency I would use the exp claim inside token, since it is the token which is passed around and the expiry inside token will be checked and this way the services would check the same thing

Its also important to check the token expiry during some processes as well - instead of passing accessToken directly to the process a method should be passed which upon calling returns the accessToken. This would ensure that token wouldn't get expired during some long running tasks